File: cross-platform/windows-controller/src-tauri/src/emulator.rs
use crate::{run_command, AvdInfo};

/// List all configured AVDs and check which are currently running.
pub async fn list_avds(emulator_bin: &str, adb: &str) -> Result<Vec<AvdInfo>, String> {
    // Get list of AVD names
    let output = run_command(emulator_bin, &["-list-avds"]).await?;
    let avd_names: Vec<String> = output
        .lines()
        .filter(|l| !l.trim().is_empty())
        .map(|l| l.trim().to_string())
        .collect();

    // Get running emulators to match serial <-> AVD name
    let running = get_running_emulators(adb).await;

    // Determine .android AVD directory
    let avd_home = std::env::var("ANDROID_AVD_HOME").unwrap_or_else(|_| {
        let home = std::env::var("USERPROFILE").unwrap_or_default();
        format!("{}\\.android\\avd", home)
    });

    let mut avds = Vec::new();
    for name in avd_names {
        let running_serial = running.iter()
            .find(|(_, avd_name)| avd_name == &name)
            .map(|(serial, _)| serial.clone());

        // Try to parse config.ini for richer info
        let config_path = format!("{}\\{}.avd\\config.ini", avd_home, name);
        let (device, target, api_level, abi, path) = parse_avd_config(&config_path, &avd_home, &name);

        avds.push(AvdInfo {
            name: name.clone(),
            device,
            path,
            target,
            api_level,
            abi,
            is_running: running_serial.is_some(),
            running_serial,
        });
    }

    Ok(avds)
}

/// Parse an AVD config.ini file for device info.
fn parse_avd_config(config_path: &str, avd_home: &str, name: &str) -> (String, String, u32, String, String) {
    let mut device = String::new();
    let mut target = String::new();
    let mut api_level: u32 = 0;
    let mut abi = String::new();
    let mut path = format!("{}\\{}.avd", avd_home, name);

    if let Ok(content) = std::fs::read_to_string(config_path) {
        for line in content.lines() {
            let line = line.trim();
            if let Some((key, value)) = line.split_once('=') {
                let key = key.trim();
                let value = value.trim();
                match key {
                    "hw.device.name" => device = value.to_string(),
                    "tag.display" | "tag.id" => {
                        if target.is_empty() {
                            target = value.to_string();
                        }
                    }
                    "image.sysdir.1" => {
                        // Extract API level from path like "system-images/android-34/google_apis/x86_64/"
                        if let Some(android_part) = value.split('/').find(|p| p.starts_with("android-")) {
                            if let Some(level_str) = android_part.strip_prefix("android-") {
                                api_level = level_str.parse().unwrap_or(0);
                            }
                        }
                        // Extract ABI from path
                        if let Some(last) = value.trim_end_matches('/').rsplit('/').next() {
                            abi = last.to_string();
                        }
                    }
                    "abi.type" => {
                        if abi.is_empty() {
                            abi = value.to_string();
                        }
                    }
                    "AvdId" | "avd.ini.displayname" => {
                        // Could use for display name but we already have `name`
                    }
                    _ => {}
                }
            }
        }
    }

    // Also try the .ini file (not config.ini) for the target
    let ini_path = format!("{}\\{}.ini", avd_home, name);
    if let Ok(content) = std::fs::read_to_string(ini_path) {
        for line in content.lines() {
            if let Some((key, value)) = line.split_once('=') {
                let key = key.trim();
                let value = value.trim();
                if key == "target" && target.is_empty() {
                    target = value.to_string();
                }
                if key == "path" && !value.is_empty() {
                    path = value.to_string();
                }
            }
        }
    }

    (device, target, api_level, abi, path)
}

/// Start an AVD emulator. Returns immediately (emulator launches in background).
pub async fn start_avd(emulator_bin: &str, name: &str, cold_boot: bool) -> Result<String, String> {
    let mut args = vec!["-avd", name, "-no-snapshot-save"];
    if cold_boot {
        args.push("-no-snapshot-load");
    }

    // Spawn detached — we don't want to block waiting for the emulator
    let child = tokio::process::Command::new(emulator_bin)
        .args(&args)
        .stdout(std::process::Stdio::null())
        .stderr(std::process::Stdio::null())
        .spawn()
        .map_err(|e| format!("Failed to start emulator: {}", e))?;

    Ok(format!("Emulator '{}' starting (PID: {:?})", name, child.id()))
}

/// Stop an emulator by sending `emu kill` via its serial.
pub async fn stop_avd(adb: &str, serial: &str) -> Result<(), String> {
    run_command(adb, &["-s", serial, "emu", "kill"]).await?;
    Ok(())
}

/// Get running emulator serials and their AVD names.
async fn get_running_emulators(adb: &str) -> Vec<(String, String)> {
    let devices_output = run_command(adb, &["devices"]).await.unwrap_or_default();
    let mut results = Vec::new();

    for line in devices_output.lines().skip(1) {
        let parts: Vec<&str> = line.split_whitespace().collect();
        if parts.len() >= 2 && parts[0].starts_with("emulator-") && parts[1] == "device" {
            let serial = parts[0].to_string();
            // Query the AVD name from the running emulator
            if let Ok(name) = run_command(adb, &["-s", &serial, "emu", "avd", "name"]).await {
                let avd_name = name.lines().next().unwrap_or("").trim().to_string();
                if !avd_name.is_empty() {
                    results.push((serial, avd_name));
                }
            }
        }
    }

    results
}

File: cross-platform/windows-controller/src-electron/services/emulator.ts
// Emulator service — wraps Android emulator CLI commands

import { execFile, spawn } from "child_process";
import { promisify } from "util";
import { existsSync, readFileSync } from "fs";
import path from "path";
import type { AvdInfo } from "./types";
import { runAdbCommand } from "./adb";

const execFileAsync = promisify(execFile);
const TIMEOUT = 15_000;
const MAX_BUFFER = 20 * 1024 * 1024;

type ExecError = Error & {
  code?: string | number;
  stderr?: string | Buffer;
  stdout?: string | Buffer;
};

function toText(value: unknown): string {
  if (Buffer.isBuffer(value)) return value.toString("utf-8");
  return typeof value === "string" ? value : String(value ?? "");
}

function formatEmulatorError(emulatorPath: string, args: string[], error: unknown): Error {
  const err = error as ExecError;
  const detail = [err.message, toText(err.stderr), toText(err.stdout)]
    .filter(Boolean)
    .join("\n")
    .split(/\r?\n/)
    .map((l) => l.trim())
    .filter(Boolean)
    .slice(0, 4)
    .join(" | ");

  if (err.code === "ENOENT") {
    return new Error(
      `Emulator executable not found: ${emulatorPath}. Configure Android SDK path in Settings.`
    );
  }

  return new Error(
    `Emulator command failed (${args.join(" ")}). ${detail || "Unknown error"}`
  );
}

async function runEmulatorCommand(
  emulatorPath: string,
  args: string[],
  timeout: number = TIMEOUT
): Promise<{ stdout: string; stderr: string }> {
  try {
    const { stdout, stderr } = await execFileAsync(emulatorPath, args, {
      timeout,
      windowsHide: true,
      maxBuffer: MAX_BUFFER,
    });
    return {
      stdout: toText(stdout),
      stderr: toText(stderr),
    };
  } catch (error) {
    throw formatEmulatorError(emulatorPath, args, error);
  }
}

export interface EmulatorDiagnosticsResult {
  emulatorPath: string;
  version: string;
  avdCount: number;
  avdNames: string[];
  details: string[];
}

export async function getEmulatorDiagnostics(
  emulatorPath: string
): Promise<EmulatorDiagnosticsResult> {
  const details: string[] = [];

  const { stdout: versionOut } = await runEmulatorCommand(emulatorPath, ["-version"], 5000);
  const versionLine = versionOut
    .split(/\r?\n/)
    .map((line) => line.trim())
    .find((line) => /^android emulator version/i.test(line));
  const version = versionLine || versionOut.split(/\r?\n/)[0]?.trim() || "Unknown";
  details.push(`Emulator version: ${version}`);

  const { stdout: avdOut } = await runEmulatorCommand(emulatorPath, ["-list-avds"]);
  const avdNames = avdOut
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter(Boolean);
  details.push(`Detected AVDs: ${avdNames.length}`);

  return {
    emulatorPath,
    version,
    avdCount: avdNames.length,
    avdNames,
    details,
  };
}

// ---------------------------------------------------------------------------
// List AVDs
// ---------------------------------------------------------------------------

export async function listAvds(
  emulatorPath: string,
  adbPath: string
): Promise<AvdInfo[]> {
  let stdout: string;
  try {
    const result = await execFileAsync(emulatorPath, ["-list-avds"], {
      timeout: TIMEOUT,
      windowsHide: true,
    });
    stdout = result.stdout;
  } catch (error) {
    throw formatEmulatorError(emulatorPath, ["-list-avds"], error);
  }

  const names = stdout
    .split("\n")
    .map((l) => l.trim())
    .filter(Boolean);

  const running = await getRunningEmulators(adbPath);

  const avdHome =
    process.env.ANDROID_AVD_HOME ||
    path.join(process.env.USERPROFILE || "", ".android", "avd");

  const avds: AvdInfo[] = [];

  for (const name of names) {
    const configPath = path.join(avdHome, `${name}.avd`, "config.ini");
    const { device, target, apiLevel, abi, avdPath } = parseAvdConfig(
      configPath,
      avdHome,
      name
    );

    const match = running.find(([, n]) => n === name);

    avds.push({
      name,
      device,
      target,
      apiLevel,
      abi,
      path: avdPath,
      running: !!match,
      serial: match ? match[0] : null,
    });
  }

  return avds;
}

// ---------------------------------------------------------------------------
// Parse AVD config
// ---------------------------------------------------------------------------

interface AvdConfig {
  device: string;
  target: string;
  apiLevel: number;
  abi: string;
  avdPath: string;
}

function parseAvdConfig(
  configPath: string,
  avdHome: string,
  name: string
): AvdConfig {
  let device = "";
  let target = "";
  let apiLevel = 0;
  let abi = "";
  let avdPath = path.join(avdHome, `${name}.avd`);

  try {
    const content = readFileSync(configPath, "utf-8");
    for (const line of content.split("\n")) {
      const trimmed = line.trim();
      const eqIdx = trimmed.indexOf("=");
      if (eqIdx < 0) continue;
      const key = trimmed.slice(0, eqIdx).trim();
      const value = trimmed.slice(eqIdx + 1).trim();

      switch (key) {
        case "hw.device.name":
          device = value;
          break;
        case "tag.display":
        case "tag.id":
          if (!target) target = value;
          break;
        case "image.sysdir.1": {
          const androidPart = value
            .split(/[\\/]/)
            .find((p) => p.startsWith("android-"));
          if (androidPart) {
            const level = parseInt(androidPart.replace("android-", ""), 10);
            if (!isNaN(level)) apiLevel = level;
          }
          const lastSegment = value.replace(/[\\/]+$/, "").split(/[\\/]/).pop();
          if (lastSegment) abi = lastSegment;
          break;
        }
        case "abi.type":
          if (!abi) abi = value;
          break;
      }
    }
  } catch {
    // config.ini may not exist
  }

  // Also check the top-level .ini file
  const iniPath = path.join(avdHome, `${name}.ini`);
  try {
    const content = readFileSync(iniPath, "utf-8");
    for (const line of content.split("\n")) {
      const eqIdx = line.indexOf("=");
      if (eqIdx < 0) continue;
      const key = line.slice(0, eqIdx).trim();
      const value = line.slice(eqIdx + 1).trim();
      if (key === "target" && !target) target = value;
      if (key === "path" && value) avdPath = value;
    }
  } catch {
    // .ini may not exist
  }

  return { device, target, apiLevel, abi, avdPath };
}

// ---------------------------------------------------------------------------
// Start / Stop AVD
// ---------------------------------------------------------------------------

export function startAvd(
  emulatorPath: string,
  name: string,
  coldBoot: boolean
): void {
  if (!existsSync(emulatorPath)) {
    throw new Error(
      `Emulator executable not found: ${emulatorPath}. Configure Android SDK path in Settings.`
    );
  }

  const args = [`@${name}`];
  if (coldBoot) args.push("-no-snapshot-load");

  // Spawn detached — emulator runs independently
  const child = spawn(emulatorPath, args, {
    detached: true,
    stdio: "ignore",
  });
  child.on("error", (err) => {
    console.error(`Failed to start emulator "${name}":`, err);
  });
  child.unref();
}

export async function stopAvd(
  adbPath: string,
  serial: string
): Promise<string> {
  const { stdout } = await runAdbCommand(adbPath, ["-s", serial, "emu", "kill"], TIMEOUT);
  return stdout.trim();
}

// ---------------------------------------------------------------------------
// Running emulators
// ---------------------------------------------------------------------------

async function getRunningEmulators(
  adbPath: string
): Promise<[string, string][]> {
  const result: [string, string][] = [];

  try {
    const { stdout } = await runAdbCommand(adbPath, ["devices"], TIMEOUT);
    const serials = stdout
      .split("\n")
      .filter((l) => l.includes("device") && !l.startsWith("List"))
      .map((l) => l.split(/\s+/)[0])
      .filter((s) => s.startsWith("emulator-"));

    for (const serial of serials) {
      try {
        const { stdout: avdName } = await runAdbCommand(
          adbPath,
          ["-s", serial, "emu", "avd", "name"],
          5000
        );
        const name = avdName.split("\n")[0].trim();
        if (name) result.push([serial, name]);
      } catch {
        // Skip if can't get name
      }
    }
  } catch {
    // ADB not available
  }

  return result;
}
